tests/pipeline_with_optional_test.py

```diff
@@ -1,229 +1,100 @@
-        sl.Pipeline([make_optional])
+    pl = sl.Pipeline([make_optional])
+    assert pl.compute(Optional[int]) == 3
 
 
-def test_provider_returning_union_disallowed() -> None:
+def test_provider_returning_union_allowed() -> None:
     def make_union() -> Union[int, float]:
         return 3
 
-    with pytest.raises(ValueError):
-        sl.Pipeline([make_union])
+    pl = sl.Pipeline([make_union])
+    assert pl.compute(Union[int, float]) == 3
 
 
-def test_parameter_type_union_or_optional_disallowed() -> None:
+def test_parameter_type_union_or_optional_allowed() -> None:
     pipeline = sl.Pipeline()
-    with pytest.raises(ValueError):
-        pipeline[Union[int, float]] = 3  # type: ignore[index]
-    with pytest.raises(ValueError):
-        pipeline[Optional[int]] = 3  # type: ignore[index]
+    pipeline[Union[int, float]] = 3  # type: ignore[index]
+    assert pipeline.compute(Union[int, float]) == 3
+    pipeline[Optional[int]] = 4  # type: ignore[index]
+    assert pipeline.compute(Optional[int]) == 4
 
 
-def test_union_requirement_satisfied_by_unique_match() -> None:
+def test_union_requirement_not_satisfied_by_any_of_its_arguments() -> None:
     def require_union(x: Union[int, float]) -> str:
         return f'{x}'
 
     pipeline = sl.Pipeline([require_union])
     pipeline[int] = 1
-    assert pipeline.compute(str) == '1'
-
-
-def test_union_requirement_with_multiple_matches_raises_AmbiguousProvider() -> None:
-    def require_union(x: Union[int, float]) -> str:
-        return f'{x}'
+    with pytest.raises(sl.UnsatisfiedRequirement):
+        pipeline.get(str)
 
     pipeline = sl.Pipeline([require_union])
-    pipeline[int] = 1
-    pipeline[float] = 1.5
-    with pytest.raises(sl.AmbiguousProvider):
-        pipeline.compute(str)
+    pipeline[float] = 1.2
+    with pytest.raises(sl.UnsatisfiedRequirement):
+        pipeline.get(str)
 
 
-def test_optional_dependency_can_be_filled_by_non_optional_param(
-    scheduler: sl.scheduler.Scheduler,
-) -> None:
+def test_optional_dependency_cannot_be_filled_by_non_optional_param() -> None:
     def use_optional(x: Optional[int]) -> str:
         return f'{x or 123}'
 
     pipeline = sl.Pipeline([use_optional], params={int: 1})
-    assert pipeline.compute(str, scheduler=scheduler) == '1'
+    with pytest.raises(sl.UnsatisfiedRequirement):
+        pipeline.get(str)
 
 
-def test_optional_dependency_can_be_filled_by_non_optional_param_kwarg(
-    scheduler: sl.scheduler.Scheduler,
-) -> None:
+def test_optional_dependency_cannot_be_filled_by_non_optional_param_kwarg() -> None:
     def use_optional(*, x: Optional[int]) -> str:
         return f'{x or 123}'
 
     pipeline = sl.Pipeline([use_optional], params={int: 1})
-    assert pipeline.compute(str, scheduler=scheduler) == '1'
+    with pytest.raises(sl.UnsatisfiedRequirement):
+        pipeline.get(str)
 
 
-def test_union_with_none_can_be_used_instead_of_Optional() -> None:
+def test_Union_and_optional_are_distinct() -> None:
```

Comment:
I don't think `Union[T, None]` and `Optional[T]` should be treated differently.
But this test doesn't actually seem to have anything to do with `Optional`. It just tests different argument orders.

Reply:
> I don't think Union[T, None] and Optional[T] should be treated differently.

Then we would need to add special-case handling.

Reply:
Ah, now I think I understand what you want to test here. But there are no optionals in the test, only unions.